File: rudder-client-ios/RudderClient/Event/RudderEventBuilder.swift
//
//  RudderEventBuilder.swift
//  RudderSample
//
//  Created by Arnab Pal on 04/07/19.
//  Copyright © 2019 Arnab Pal. All rights reserved.
//

import Foundation

class RudderEventBuilder {
    internal var property: RudderProperty = RudderProperty()
    func setPropertyBuilder(builder: RudderPropertyBuilder) throws -> RudderEventBuilder {
        self.property = try builder.build()
        return self
    }
    
    func setProperty(property: RudderProperty) -> RudderEventBuilder {
        self.property = property
        return self
    }
    
    internal var event: String = ""
    func setEvent(event: String) -> RudderEventBuilder {
        self.event = event
        return self
    }
    
    internal var channel: String = ""
    func setChannel(channel: String) -> RudderEventBuilder {
        self.channel = channel
        return self
    }
    
    func build() -> RudderEvent {
        return RudderEvent().fromBuilder(builder: self, template: EventRepository.eventTemplate!)
    }
}

File: RudderUnitySdkProj/Assets/rudder-sdk/Event/RudderEventBuilder.cs
using com.rudderlabs.unity.library.Event.Property;

namespace com.rudderlabs.unity.library.Event
{
    public class RudderEventBuilder
    {
        private RudderProperty rudderProperty;
        public RudderEventBuilder SetRudderProperty(RudderProperty rudderProperty)
        {
            this.rudderProperty = rudderProperty;
            return this;
        }

        public RudderEventBuilder SetRudderProperty(RudderPropertyBuilder propertyBuilder)
        {
            this.rudderProperty = propertyBuilder.Build();
            return this;
        }

        private string eventName;
        public RudderEventBuilder SetEventName(string eventName)
        {
            this.eventName = eventName;
            return this;
        }

        private string userId;
        public RudderEventBuilder SetUseId(string userId)
        {
            this.userId = userId;
            return this;
        }

        private string channel;
        public RudderEventBuilder SetChannel(string channel)
        {
            this.channel = channel;
            return this;
        }

        private RudderUserProperty userProperty;
        public RudderEventBuilder SetUserProperty(RudderUserProperty userProperty)
        {
            this.userProperty = userProperty;
            return this;
        }

        public RudderEvent Build()
        {
            RudderEvent rudderEvent = new RudderEvent();
            if (this.rudderProperty != null)
            {
                rudderEvent.SetProperty(this.rudderProperty);
            }
            if (this.userProperty != null)
            {
                rudderEvent.SetUserProperty(this.userProperty);
            }
            rudderEvent.message.eventName = eventName;
            rudderEvent.message.userId = userId;
            rudderEvent.message.channel = channel;
            return rudderEvent;
        }
    }
}
